include/openmc/random_ray/random_ray.h

```diff
@@ -0,0 +1,44 @@
+#ifndef OPENMC_RANDOM_RAY_RAY_H
+#define OPENMC_RANDOM_RAY_RAY_H
+
+#include "openmc/particle.h"
+
+namespace openmc {
+
+/*
+ * The RandomRay class encompasses data and methods for transporting random rays
+ * through the model. It is a small extension of the Particle class.
+ */
+
+class RandomRay : public Particle {
```

Comment:
Can't this  just inherit from `GeometryState` now?

Reply:
I was hoping to do that, but when I tried it turned out there are a unfortunately still a couple of less obvious areas throughout the geometry code that rely on a `Particle` rather than a `GeometryState`. It would likely make for a good separate PR to finish abstracting everything practical into the new `GeometryState`.

Reply:
Is it really that many? It seems like the wgt() variable and the boundary conditions are the only spots.

Reply:
Here's what happens when I try it:

```
/home/jtramm/openmc/src/boundary_condition.cpp: In member function ‘virtual void openmc::VacuumBC::handle_particle(openmc::Particle&, const openmc::Surface&) const’:
/home/jtramm/openmc/src/boundary_condition.cpp:26:20: error: invalid ‘static_cast’ from type ‘openmc::Particle*’ to type ‘openmc::RandomRay*’
   26 |     RandomRay* r = static_cast<RandomRay*>(&p);
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/libopenmc.dir/build.make:90: CMakeFiles/libopenmc.dir/src/boundary_condition.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/jtramm/openmc/src/random_ray/random_ray.cpp: In member function ‘uint64_t openmc::RandomRay::transport_history_based_single_ray()’:
/home/jtramm/openmc/src/random_ray/random_ray.cpp:86:10: error: ‘alive’ was not declared in this scope
   86 |   while (alive()) {
      |          ^~~~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp:90:5: error: ‘event_cross_surface’ was not declared in this scope
   90 |     event_cross_surface();
      |     ^~~~~~~~~~~~~~~~~~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp:93:10: error: ‘n_event’ was not declared in this scope
   93 |   return n_event();
      |          ^~~~~~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp: In member function ‘void openmc::RandomRay::event_advance_ray()’:
/home/jtramm/openmc/src/random_ray/random_ray.cpp:114:7: error: ‘wgt’ was not declared in this scope
  114 |       wgt() = 0.0;
      |       ^~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp:134:9: error: ‘wgt’ was not declared in this scope
  134 |         wgt() = 0.0;
      |         ^~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp: In member function ‘void openmc::RandomRay::attenuate_flux(double, bool)’:
/home/jtramm/openmc/src/random_ray/random_ray.cpp:167:3: error: ‘n_event’ was not declared in this scope
  167 |   n_event()++;
      |   ^~~~~~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp: In member function ‘void openmc::RandomRay::initialize_ray(uint64_t, int, openmc::FlatSourceDomain*)’:
/home/jtramm/openmc/src/random_ray/random_ray.cpp:241:3: error: ‘n_event’ was not declared in this scope
  241 |   n_event() = 0;
      |   ^~~~~~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp:248:3: error: ‘wgt’ was not declared in this scope
  248 |   wgt() = 1.0;
      |   ^~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp:256:38: error: ‘seeds’ was not declared in this scope
  256 |   init_particle_seeds(particle_seed, seeds());
      |                                      ^~~~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp:257:3: error: ‘stream’ was not declared in this scope; did you mean ‘std::io_errc::stream’?
  257 |   stream() = STREAM_TRACKING;
      |   ^~~~~~
      |   std::io_errc::stream
In file included from /usr/include/c++/11/ios:42,
                 from /usr/include/c++/11/ostream:38,
                 from /usr/include/c++/11/iostream:39,
                 from /home/jtramm/openmc/include/openmc/position.h:5,
                 from /home/jtramm/openmc/include/openmc/particle_data.h:6,
                 from /home/jtramm/openmc/include/openmc/particle.h:12,
                 from /home/jtramm/openmc/include/openmc/random_ray/random_ray.h:4,
                 from /home/jtramm/openmc/src/random_ray/random_ray.cpp:1:
/usr/include/c++/11/bits/ios_base.h:203:24: note: ‘std::io_errc::stream’ declared here
  203 |   enum class io_errc { stream = 1 };
      |                        ^~~~~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp:261:54: error: ‘current_seed’ was not declared in this scope
  261 |     model::external_sources[sampling_source]->sample(current_seed())};
      |                                                      ^~~~~~~~~~~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp:265:3: error: ‘from_source’ was not declared in this scope
  265 |   from_source(&site);
      |   ^~~~~~~~~~~
/home/jtramm/openmc/src/random_ray/random_ray.cpp:275:9: error: ‘cell_born’ was not declared in this scope
  275 |     if (cell_born() == C_NONE)
      |         ^~~~~~~~~
make[2]: *** [CMakeFiles/libopenmc.dir/build.make:720: CMakeFiles/libopenmc.dir/src/random_ray/random_ray.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:266: CMakeFiles/libopenmc.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

Some of these are not too bad, but others are more of a project. The `event_cross_surface()` has a lot of neutron-specific stuff in it, although the basic idea of handling the geometry updates when doing this is more suited to a `Geometron`. `from_source()` is needed by the ray as well. There are portions in that logic that would apply to any point in the geometry (e.g., location, direction), and others that are more specific to a neutron etc, like energy. The PRNG seeds are an issue as well, although that one can probably be handled pretty easily by putting that stuff in its own class, and then having the `Particle` and `RandomRay` both inherit from it. Or, as it's only used by the `from_source()` routine in random ray mode, perhaps the random ray solver could implement its own.

Anyway, there's enough issues there that it would likely be better left to future optimization work to rework things to allow for inheritance from the `Geometron` rather than the full `Particle`.

Reply:
My new plot PR fixes the event_cross_surface stuff, I'm pretty sure.

Yeah, makes sense. This would be a straightforward one for me to handle in the future! Can you mark it as an optimization to try in the future? Aside from that, it all sounds great!

Reply:
I've marked this as a "TODO" item in the source.